helix-core/src/main/java/org/apache/helix/HelixManager.java

```diff
@@ -362,6 +363,13 @@ void addExternalViewChangeListener(org.apache.helix.ExternalViewChangeListener l
    */
   String getMetadataStoreConnectionString();
 
+  /**
+   * @return the RealmAwareZkConnectionConfig usd to create a realm aware zkClient
+   */
+  default RealmAwareZkClient.RealmAwareZkConnectionConfig getRealmAwareZkConnectionConfig() {
```

Comment:
A high-level comment:

- The overarching principle here is that the underlying ZkClients in the given HelixManager and the ZkClient we are using in TaskStateModelFactory have to have the same connection config.
- Instead of passing ZkConnectionConfig that returns a null as the default implementation, can we just have them share the underlying ZkClient?
- This way, we can guarantee that whatever HelixManager sees is exactly the same as what we see in TaskStateModelFactory, and we do not have to worry about managing the lifecycles of two separate ZkClients.

Reply:
This implementation is a result of our (you and I mainly) discussion when Configurable Task Pool was first created, and I don't think the situation has changed. 

See line 137 in TaskStateModelFactory.java where the TODO is. The main problem is that some customers register TaskStateModel before connecting the manager, some do it after. For those that register before connecting, we have to create a separate zkClient. If we enforce the ordering, customers have to change participant code, which we decided against.

Reply:
Thanks for explaining the context. Yes, there are implementations that may not connect the manager prior to registering their task. Can we add a TODO or a separate issue tracking this? Now that we have different modes of underlying ZkClient instantiation, we need to be more careful about not creating multiple ZkClients, which could only add to the confusion for future developers.

For the extent of this PR, I'd be fine creating a one-off ZkClient using the connection config.

With that said, why does this need to be a default implementation? Nothing wrong with using the default interface, but in this case, the default behavior is not well-defined. I prefer not making this default.

Reply:
A TODO has been added at the place where the zkClient is created, so I think we have that covered. 

About the default behavior, none of the test-only HelixManagers have the `_helixManagerProperty` field, so even if I add implementations, they will be `return null`.

Reply:
I think we should not even add the method to the HelixManager interface. 1. it is ZK config, conceptually, HelixManager does not have to be implemented by a ZkHelixManager. 2. if we expose more internal implementation details from the public interface, we may be stuck with this design and cannot change anything in the future. This is not a good OO design.

Please see my proposal that I put here earlier.

Reply:
Resolved with @jiajunwang offline: not a good OO design, so better solution: moving this to `ZkHelixManager` only.